File: src/examples/ex_dual_arm.cpp
/* ex_dual_arm.cpp
 * Two Kinova Gen3 arms, each fitted with a Robotiq 2F-85 gripper,
 * in a shared MuJoCo scene.
 *
 * arm1 at x = -1.0 m, facing +X.
 * arm2 at x = +1.0 m, facing +X; all element names prefixed "r2_".
 *
 * Both arms hold the home pose via PD + KDL gravity compensation.
 * Grippers cycle open/closed every 3 s.
 *
 * Gravity comp uses KDL::ChainDynParam built from a chain that includes
 * the gripper's lumped inertia (via init_robot_from_mjcf tool_body).
 * jnt_trq_cmd is primed before the loop so the first physics step already
 * gets correct compensation.
 *
 * Requires MuJoCo Menagerie in cache.
 *
 * Usage:
 *   ex_dual_arm [--headless]
 *
 * Runs 600 steps and exits; --headless skips the viewer and prints both EE positions. */

#include "mj_kdl_wrapper/mj_kdl_wrapper.hpp"
#include "common.hpp"
#include "example_paths.hpp"

#include <kdl/chaindynparam.hpp>
#include <kdl/chainfksolverpos_recursive.hpp>

#include <cmath>
#include <iostream>
#include <string>

using mj_kdl_examples::kHomePose;

static constexpr double kKp[7] = { 100, 200, 100, 200, 100, 200, 100 };
static constexpr double kKd[7] = { 10, 20, 10, 20, 10, 20, 10 };
static constexpr int    kSteps = 600;

using mj_kdl_examples::kGripperClosed;

int main(int argc, char *argv[])
{
    const bool headless = mj_kdl_examples::parse_args(argc, argv).headless;

    const std::string arm_mjcf = mj_kdl_examples::menagerie_model("kinova_gen3/gen3.xml");
    const std::string grp_mjcf = mj_kdl_examples::asset("robotiq_2f85/2f85.xml");

    mj_kdl::AttachmentSpec gs;
    gs.mjcf_path = grp_mjcf.c_str();
    gs.attach_to = { mj_kdl::AttachKind::Site, "pinch_site" };
    gs.prefix    = "g_";

    mj_kdl::RobotSpec arm1_spec;
    arm1_spec.path   = arm_mjcf.c_str();
    arm1_spec.prefix = "";
    arm1_spec.pos[0] = -1.0;
    arm1_spec.attachments.push_back(gs);

    mj_kdl::RobotSpec arm2_spec;
    arm2_spec.path   = arm_mjcf.c_str();
    arm2_spec.prefix = "r2_";
    arm2_spec.pos[0] = 1.0;
    arm2_spec.attachments.push_back(gs);

    mj_kdl::SceneSpec sc;
    sc.timestep   = 0.002;
    sc.add_floor  = true;
    sc.add_skybox = true;
    sc.robots.push_back(arm1_spec);
    sc.robots.push_back(arm2_spec);

    mj_kdl::Env env;
    if (!mj_kdl::init_env(&env, &sc)) {
        std::cerr << "init_env() failed\n";
        return 1;
    }

    mj_kdl::ToolFrameSpec tool1, tool2;
    tool1.tool_body = "g_base_mount";
    tool1.tcp_site  = "g_pinch";
    tool2.tool_body = "r2_g_base_mount";
    tool2.tcp_site  = "r2_g_pinch";

    mj_kdl::Robot arm1, arm2;
    if (
      !mj_kdl::init_robot_from_mjcf(&arm1, &env, "base_link", "bracelet_link", "", &tool1)
      || !mj_kdl::init_robot_from_mjcf(
        &arm2, &env, "r2_base_link", "r2_bracelet_link", "", &tool2
      )
    ) {
        std::cerr << "init_robot_from_mjcf() failed\n";
        return 1;
    }

    const int                  n     = arm1.n_joints;
    mj_kdl::SceneActuatorSlot *fing1 = mj_kdl::bind_scene_actuator(&env.scene, "g_fingers_actuator");
    mj_kdl::SceneActuatorSlot *fing2 =
      mj_kdl::bind_scene_actuator(&env.scene, "r2_g_fingers_actuator");
    if (!fing1 || !fing2) return 1;

    KDL::ChainDynParam dyn1(arm1.chain, KDL::Vector(0.0, 0.0, -9.81));
    KDL::ChainDynParam dyn2(arm2.chain, KDL::Vector(0.0, 0.0, -9.81));

    KDL::JntArray q_home(n), q1(n), q2(n), g1(n), g2(n);
    for (int i = 0; i < n; ++i) q_home(i) = kHomePose[i];
    if (
      !mj_kdl::set_control_mode(&arm1, mj_kdl::CtrlMode::TORQUE)
      || !mj_kdl::set_control_mode(&arm2, mj_kdl::CtrlMode::TORQUE)
    )
        return 1;

    // Prime jnt_trq_cmd so the first physics step already gets gravity compensation.
    auto prime_grav = [&]() {
        dyn1.JntToGravity(q_home, g1);
        dyn2.JntToGravity(q_home, g2);
        for (int i = 0; i < n; ++i) {
            arm1.jnt_trq_cmd[i] = g1(i);
            arm2.jnt_trq_cmd[i] = g2(i);
        }
    };

    // reset() seeds the finger slots from ctrl, so the hook sets ctrl.
    env.on_reset = [&](mj_kdl::ResetContext *ctx) {
        mj_kdl::set_joint_pos(&arm1, q_home);
        mj_kdl::set_joint_pos(&arm2, q_home);
        ctx->data->ctrl[fing1->ctrl_id] = kGripperClosed;
        ctx->data->ctrl[fing2->ctrl_id] = kGripperClosed;
        prime_grav();
    };

    mj_kdl::reset(&env);
    mj_kdl::update(&env);

    // Per-step: update() reads sensors and flushes the previous jnt_trq_cmd;
    // then compute PD + KDL gravity for the next step.
    auto ctrl_step = [&]() {
        mj_kdl::update(&env);
        for (int i = 0; i < n; ++i) q1(i) = arm1.jnt_pos_msr[i];
        for (int i = 0; i < n; ++i) q2(i) = arm2.jnt_pos_msr[i];
        dyn1.JntToGravity(q1, g1);
        dyn2.JntToGravity(q2, g2);
        for (int i = 0; i < n; ++i) {
            arm1.jnt_trq_cmd[i] =
              kKp[i] * (kHomePose[i] - arm1.jnt_pos_msr[i]) - kKd[i] * arm1.jnt_vel_msr[i] + g1(i);
            arm2.jnt_trq_cmd[i] =
              kKp[i] * (kHomePose[i] - arm2.jnt_pos_msr[i]) - kKd[i] * arm2.jnt_vel_msr[i] + g2(i);
        }
        fing1->command = (std::fmod(env.data->time, 6.0) < 3.0) ? kGripperClosed : 0.0;
        fing2->command = fing1->command;
    };

    if (headless) {
        for (int step = 0; step < kSteps; ++step) {
            ctrl_step();
            mj_kdl::step(&env);
        }

        KDL::ChainFkSolverPos_recursive fk1(arm1.chain), fk2(arm2.chain);
        KDL::JntArray                   q1(n), q2(n);
        KDL::Frame                      ee1, ee2;
        for (int i = 0; i < n; ++i) {
            q1(i) = arm1.jnt_pos_msr[i];
            q2(i) = arm2.jnt_pos_msr[i];
        }
        fk1.JntToCart(q1, ee1);
        fk2.JntToCart(q2, ee2);
        std::cout << "arm1 EE: [" << ee1.p.x() << ", " << ee1.p.y() << ", " << ee1.p.z() << "]\n";
        std::cout << "arm2 EE: [" << ee2.p.x() << ", " << ee2.p.y() << ", " << ee2.p.z() << "]\n";
    } else {
        if (!mj_kdl::open_viewer(&env)) {
            std::cerr << "open_viewer() failed\n";
            return 1;
        }
        for (int step = 0; step < kSteps; ++step) {
            ctrl_step();
            if (!mj_kdl::step(&env)) break;
            mj_kdl::pace_realtime(&env);
        }
    }

    mj_kdl::cleanup(&env);
    return 0;
}

File: python/examples/ex_dual_arm.py
#!/usr/bin/env python3
"""Dual-arm example ported from src/examples/ex_dual_arm.cpp.

Two Kinova GEN3 arms, each with a Robotiq 2F-85 gripper, in one scene. Both run
joint-space impedance to the home pose. An Env on_reset hook re-homes both arms.
"""

from __future__ import annotations

import argparse
import math
import os
from pathlib import Path

import mj_kdl_wrapper as mjk

DEFAULT_MODEL = "third_party/menagerie/kinova_gen3/gen3.xml"
DEFAULT_GRIPPER = "assets/robotiq_2f85/2f85.xml"
MODEL_ENV_VAR = "MJ_KDL_MODEL"
GRIPPER_ENV_VAR = "MJ_KDL_GRIPPER"
HOME_POSE = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]

KP = [100, 200, 100, 200, 100, 200, 100]
KD = [10, 20, 10, 20, 10, 20, 10]


def path_from_arg(value: str, label: str) -> Path:
    path = Path(value)
    if not path.exists():
        raise FileNotFoundError(f"{value} does not exist for {label}")
    return path


def attachment_gripper(path: Path, prefix: str = "g_") -> mjk.AttachmentSpec:
    spec = mjk.AttachmentSpec()
    spec.mjcf_path = str(path)
    spec.attach_to = mjk.AttachTarget(mjk.AttachKind.Site, "pinch_site")
    spec.prefix = prefix
    return spec


def apply(robot: mjk.Robot, target: list[float]) -> None:
    robot.update()
    grav = robot.gravity_torques(-9.81)
    robot.jnt_trq_cmd = [
        KP[i] * (target[i] - robot.jnt_pos_msr[i]) - KD[i] * robot.jnt_vel_msr[i] + grav[i]
        for i in range(robot.n_joints)
    ]


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true")
    args = parser.parse_args()

    arm_path = path_from_arg(mjk.menagerie.model_path("kinova_gen3", env_var=MODEL_ENV_VAR), "arm model")
    gripper_path = path_from_arg(os.environ.get(GRIPPER_ENV_VAR, DEFAULT_GRIPPER), "gripper model")
    attach = attachment_gripper(gripper_path)

    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True
    left = mjk.RobotSpec()
    left.path = str(arm_path)
    left.pos = [-1.0, 0.0, 0.0]
    left.attachments = [attach]
    right = mjk.RobotSpec()
    right.path = str(arm_path)
    right.prefix = "r2_"
    right.pos = [1.0, 0.0, 0.0]
    right.attachments = [attach]
    spec.robots = [left, right]

    env = mjk.Env.build(spec)
    try:
        tool1 = mjk.ToolFrameSpec()
        tool1.tool_body = "g_base"
        tool1.tcp_site = "g_pinch"
        tool2 = mjk.ToolFrameSpec()
        tool2.tool_body = "r2_g_base"
        tool2.tcp_site = "r2_g_pinch"
        arm1 = env.create_robot("base_link", "bracelet_link", tool=tool1)
        arm2 = env.create_robot("r2_base_link", "r2_bracelet_link", tool=tool2)
        for robot in (arm1, arm2):
            robot.ctrl_mode = mjk.CtrlMode.TORQUE

        def on_reset(ctx):
            arm1.set_joint_pos(HOME_POSE, call_forward=False)
            arm2.set_joint_pos(HOME_POSE, call_forward=False)

        env.on_reset = on_reset
        env.reset()

        def step():
            apply(arm1, HOME_POSE)
            apply(arm2, HOME_POSE)
            grip = 255.0 if math.fmod(env.time(), 6.0) < 3.0 else 0.0
            for name in ("g_fingers_actuator", "r2_g_fingers_actuator"):
                if env.has_actuator(name):
                    env.set_actuator_ctrl(name, grip)

        if args.gui:
            viewer = mjk.SimulateViewer.open(arm1, "ex_dual_arm.py")
            prev = env.time()
            try:
                while viewer.is_running():
                    if env.time() < prev - 1e-6:
                        env.reset()
                    prev = env.time()
                    step()
                    if not viewer.step():
                        break
            finally:
                viewer.close()
        else:
            end = env.time() + 1.2
            while env.time() < end:
                step()
                arm1.step()
            arm1_frame = arm1.fk_frame()
            arm2_frame = arm2.fk_frame()
            arm1_pos = [arm1_frame.p.x(), arm1_frame.p.y(), arm1_frame.p.z()]
            arm2_pos = [arm2_frame.p.x(), arm2_frame.p.y(), arm2_frame.p.z()]
            print(f"arm1 EE: {[round(x, 4) for x in arm1_pos]}")
            print(f"arm2 EE: {[round(x, 4) for x in arm2_pos]}")
    finally:
        env.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
